examples/contrastive/contrastive.cpp

```diff
@@ -0,0 +1,229 @@
+        params_amateur.prompt = "Hello my name is";
+    }
+
+    // total length of the sequence including the prompt
+    const int n_len = 32;
+
+    // init LLM
+
+    llama_backend_init(params_expert.numa);
+
+    // initialize the model
+
+    llama_model_params model_params = llama_model_default_params();
+
+    // model_params.n_gpu_layers = 99; // offload all layers to the GPU
+
+    llama_model * model_expert = llama_load_model_from_file(params_expert.model.c_str(), model_params);
+    llama_model * model_amateur = llama_load_model_from_file(params_amateur.model.c_str(), model_params);
+
+
+    if (model_expert == NULL) {
+        fprintf(stderr, "%s: error: unable to load expert model\n", __func__);
+        return 1;
+    }
+
+    if (model_amateur == NULL) {
+        fprintf(stderr, "%s: error: unable to load amateur model\n", __func__);
+        return 1;
+    }
+
+    // initialize the context
+
+    llama_context_params ctx_params = llama_context_default_params();
+
+    ctx_params.n_ctx = 2048;
+    ctx_params.n_threads = params_expert.n_threads;
+    ctx_params.n_threads_batch = params_expert.n_threads_batch == -1 ? params_expert.n_threads : params_expert.n_threads_batch;
+
+    llama_context * ctx_expert = llama_new_context_with_model(model_expert, ctx_params);
+    llama_context * ctx_amateur = llama_new_context_with_model(model_amateur, ctx_params);
+
+    if (ctx_expert == NULL) {
+        fprintf(stderr, "%s: error: failed to create the llama_context for expert\n", __func__);
+        return 1;
+    }
+
+    if (ctx_amateur == NULL) {
+        fprintf(stderr, "%s: error: failed to create the llama_context for amateur\n", __func__);
+        return 1;
+    }
+
+    // tokenize the prompt
+
+    std::vector<llama_token> tokens_list;
+    tokens_list = ::llama_tokenize(ctx_expert, params_expert.prompt, true);
+
+    const int n_ctx    = llama_n_ctx(ctx_expert);
+    const int n_kv_req = tokens_list.size() + (n_len - tokens_list.size());
+
+    LOG_TEE("\n%s: n_len = %d, n_ctx = %d, n_kv_req = %d\n", __func__, n_len, n_ctx, n_kv_req);
+
+    // make sure the KV cache is big enough to hold all the prompt and generated tokens
+    if (n_kv_req > n_ctx) {
+        LOG_TEE("%s: error: n_kv_req > n_ctx, the required KV cache size is not big enough\n", __func__);
+        LOG_TEE("%s:        either reduce n_parallel or increase n_ctx\n", __func__);
+        return 1;
+    }
+
+    // print the prompt token-by-token
+
+    fprintf(stderr, "\n");
+
+    for (auto id : tokens_list) {
+        fprintf(stderr, "%s", llama_token_to_piece(ctx_expert, id).c_str());
+    }
+
+    // create a llama_batch with size 512
+    // we use this object to submit token data for decoding
+
+    llama_batch batch = llama_batch_init(512, 0, 1);
+
+    // evaluate the initial prompt
+    for (size_t i = 0; i < tokens_list.size(); i++) {
+        llama_batch_add(batch, tokens_list[i], i, { 0 }, false);
+    }
```

Comment:
We need to check that tokens_list.size() <= batch.n_tokens here so we don't segfault if the prompt is too long.